Adds internal timezone_from_offset function

Tpt opened the pull request:
Adds internal timezone_from_offset function

It allows to build conversions from chrono without direct access to the C API

Tpt commented:
This change is a first step for #3633. It allows to remove all C API calls from the `chrono` conversion.

I am not sure adding two functions is the best way to go. An other option might be to convert the three functions constructing `datetime.timezone` to associated functions of the `PyTzInfo` object.

adamreichold commented on src/types/datetime.rs:
```diff
@@ -487,6 +488,32 @@ pub fn timezone_utc(py: Python<'_>) -> &PyTzInfo {
+/// Equivalent to `datetime.timezone` constructor
+///
+/// Use [`timezone_from_offset_and_name`] for a named timezones.
+pub fn timezone_from_offset<'a>(py: Python<'a>, offset: &PyDelta) -> PyResult<&'a PyTzInfo> {
```
My main question here is why this has to become new public API if it is mainly meant as extension point to differentiate between accessing unlimited and limited API? Wouldn't `pub(crate)` suffice for that purpose?

Tpt replied:
Yes, indeed. Done.

davidhewitt commented:
Note that there is a long history of why these functions don't exist, see https://github.com/PyO3/pyo3/pull/1588#issuecomment-1174614451

I'm not totally against adding them, but also do understand why others did not want them.

Tpt commented:
> Note that there is a long history of why these functions don't exist, see [#1588 (comment)](https://github.com/PyO3/pyo3/pull/1588#issuecomment-1174614451)
> 
> I'm not totally against adding them, but also do understand why others did not want them.

Thank you! I was not aware of this discussion. The concerns make sense. I have rewritten this MR to make `timezone_from_offset` private and I dropped `timezone_from_offset_and_name`.

davidhewitt commented on newsfragments/3648.changed.md:
No need to document an internal refactor; I've added the skip-changelog label. Please remove this fragment 😊

Tpt replied:
Done.

davidhewitt reviewed (approved):
Overall this now looks like it just relocated the function, but I've no problem with that if it's a helpful first step for the follow up feature work 👍
